fuse_models/include/fuse_models/unicycle_2d_state_cost_function.h

```diff
@@ -115,41 +113,46 @@ class Unicycle2DStateCostFunction : public ceres::SizedCostFunction<8, 2, 1, 2,
    *                         computed for the parameters where jacobians[i] is not NULL.
    * @return The return value indicates whether the computation of the residuals and/or jacobians was successful or not.
    */
-  bool Evaluate(double const* const* parameters,
-                double* residuals,
-                double** jacobians) const override
+  bool Evaluate(double const* const* parameters, double* residuals, double** jacobians) const override
   {
-    double position_pred_x;
-    double position_pred_y;
-    double yaw_pred;
+    double delta_position_pred_x;
+    double delta_position_pred_y;
+    double delta_yaw_pred;
     double vel_linear_pred_x;
     double vel_linear_pred_y;
     double vel_yaw_pred;
     double acc_linear_pred_x;
     double acc_linear_pred_y;
     predict(
-      parameters[0][0],  // position1_x
-      parameters[0][1],  // position1_y
-      parameters[1][0],  // yaw1
+      0.0,
+      0.0,
+      0.0,
       parameters[2][0],  // vel_linear1_x
       parameters[2][1],  // vel_linear1_y
       parameters[3][0],  // vel_yaw1
       parameters[4][0],  // acc_linear1_x
       parameters[4][1],  // acc_linear1_y
       dt_,
-      position_pred_x,
-      position_pred_y,
-      yaw_pred,
+      delta_position_pred_x,
+      delta_position_pred_y,
+      delta_yaw_pred,
       vel_linear_pred_x,
       vel_linear_pred_y,
       vel_yaw_pred,
       acc_linear_pred_x,
       acc_linear_pred_y,
       jacobians);
 
-    residuals[0] = parameters[5][0] - position_pred_x;
-    residuals[1] = parameters[5][1] - position_pred_y;
-    residuals[2] = parameters[6][0] - yaw_pred;
+    const double delta_yaw_est = parameters[6][0] - parameters[1][0];
+    const fuse_core::Vector2d position1(parameters[0][0], parameters[0][1]);
+    const fuse_core::Vector2d position2(parameters[5][0], parameters[5][1]);
+    const fuse_core::Vector2d position_diff = (position2 - position1);
+    const double sin_yaw_inv = std::sin(-parameters[1][0]);
+    const double cos_yaw_inv = std::cos(-parameters[1][0]);
+
+    residuals[0] = (cos_yaw_inv * position_diff.x() - sin_yaw_inv * position_diff.y()) - delta_position_pred_x;
+    residuals[1] = (sin_yaw_inv * position_diff.x() + cos_yaw_inv * position_diff.y()) - delta_position_pred_y;
```

Comment:
I think we should declare `R_yaw_inv` here and use it to compute the first two residuals:
```suggestion
    const fuse_core::Vector2d delta_position_delta(delta_position_delta_x, delta_position_delta_y);
    const fuse_core::Matrix2d R_yaw_inv = fuse_core::rotationMatrix2D(-parameters[1][0]);

    residuals.head<2>() = R_yaw_inv * position_diff - delta_position_delta;
```

I think you could even declare `delta_position_delta` before calling `predict` and pass `delta_position_delta.x()` and `delta_position_delta.y()` as arguments, so you avoid declaring `delta_position_delta_x` and `delta_position_delta_y`. I'm not too sure about that though.